### PR TITLE
[core][metric] Add worker heap memory metric export 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

This PR adds the exporting of memory used by the in memory worker memory store, which includes objects like: 
- error placeholders for objects stored in plasma 
- small returned objects inlined

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/store_provider/memory_store/memory_store.h [hunk 1]
```diff
@@ -226,9 +229,10 @@ class CoreWorkerMemoryStore {
   int32_t num_in_plasma_ GUARDED_BY(mu_) = 0;
   /// Number of objects that don't exist in the plasma store.
   int32_t num_local_objects_ GUARDED_BY(mu_) = 0;
-  /// Number of object store memory used by this memory store. (It doesn't include plasma
-  /// store memory usage).
-  int64_t used_object_store_memory_ GUARDED_BY(mu_) = 0;
```
Comment (rickyyx) [inline src/ray/core_worker/store_provider/memory_store/memory_store.h:231]: The original naming was confusing to me

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -839,6 +839,8 @@ void CoreWorker::RecordMetrics() {
   task_manager_->RecordMetrics();
   // Record metrics for executed tasks.
   task_counter_.RecordMetrics();
+  // Record worker heap memory metric
```
Comment (ericl) [inline src/ray/core_worker/core_worker.cc:842]:
```suggestion
  // Record worker heap memory metrics.
```

**Review by ericl — APPROVED** [review]

File: src/ray/stats/metric_defs.cc [hunk 3]
```diff
@@ -234,6 +234,9 @@ DEFINE_stats(object_store_memory,
              ///    - MMAP_SHM: currently in shared memory(e.g. /dev/shm).
              ///    - MMAP_DISK: memory that's fallback allocated on mmapped disk,
              ///      e.g. /tmp.
+             ///    - WORKER_HEAP: memory incurred due to inline objects, i.e.
```
Comment (rkooo567) [inline src/ray/stats/metric_defs.cc:237]: Can we make it a bit more user-friendly? 

```
Ray objects are stored in process memory. It is used for storing small objects that are smaller than 100KB. 
```

**Review by rkooo567 — APPROVED** [review]

**Comment by ericl** [comment]
CPP worker test failed, is it relevant?

**Comment by rickyyx** [comment]
I don't think so - it should be fixed by this: https://github.com/ray-project/ray/commit/881ff174a50e887ff4cd42ca68ec516046170669

Let me rebase

**Pull request merged**